Flaky test: 6/10k data race in TestAccountCheckWindowSizeWithSmallWindow

https://github.com/grpc/grpc-go/runs/5744477531?check_suite_focus=true

```
==================
WARNING: DATA RACE
Read at 0x00c0001ae930 by goroutine 160:
  runtime.mapaccess1_fast32()
      /opt/hostedtoolcache/go/1.18.0/x64/src/runtime/map_fast32.go:13 +0x0
  google.golang.org/grpc/internal/transport.testFlowControlAccountCheck()
      /home/runner/work/grpc-go/grpc-go/internal/transport/transport_test.go:1571 +0xc06
  google.golang.org/grpc/internal/transport.s.TestAccountCheckWindowSizeWithSmallWindow()
      /home/runner/work/grpc-go/grpc-go/internal/transport/transport_test.go:1478 +0x55
  google.golang.org/grpc/internal/transport.(*s).TestAccountCheckWindowSizeWithSmallWindow()
      <autogenerated>:1 +0x1d
  runtime.call16()
      /opt/hostedtoolcache/go/1.18.0/x64/src/runtime/asm_amd64.s:701 +0x48
  reflect.callMethod()
      <autogenerated>:1 +0x64
  reflect.methodValueCall()
      /opt/hostedtoolcache/go/1.18.0/x64/src/reflect/asm_amd64.s:76 +0x79
  testing.tRunner()
      /opt/hostedtoolcache/go/1.18.0/x64/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.18.0/x64/src/testing/testing.go:1486 +0x47

Previous write at 0x00c0001ae930 by goroutine 190:
  runtime.mapdelete_fast32()
      /opt/hostedtoolcache/go/1.18.0/x64/src/runtime/map_fast32.go:273 +0x0
  google.golang.org/grpc/internal/transport.(*loopyWriter).cleanupStreamHandler()
      /home/runner/work/grpc-go/grpc-go/internal/transport/controlbuf.go:757 +0x145
  google.golang.org/grpc/internal/transport.(*loopyWriter).handle()
      /home/runner/work/grpc-go/grpc-go/internal/transport/controlbuf.go:829 +0x272
  google.golang.org/grpc/internal/transport.(*loopyWriter).run()
      /home/runner/work/grpc-go/grpc-go/internal/transport/controlbuf.go:538 +0x13b
  google.golang.org/grpc/internal/transport.newHTTP2Client.func3()
      /home/runner/work/grpc-go/grpc-go/internal/transport/http2_client.go:415 +0x105

Goroutine 160 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.18.0/x64/src/testing/testing.go:1486 +0x724
  google.golang.org/grpc/internal/grpctest.RunSubTests()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/grpctest.go:99 +0x507
  google.golang.org/grpc/internal/transport.Test()
      /home/runner/work/grpc-go/grpc-go/internal/transport/transport_test.go:57 +0x3c
  testing.tRunner()
      /opt/hostedtoolcache/go/1.18.0/x64/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.18.0/x64/src/testing/testing.go:1486 +0x47

Goroutine 190 (running) created at:
  google.golang.org/grpc/internal/transport.newHTTP2Client()
      /home/runner/work/grpc-go/grpc-go/internal/transport/http2_client.go:[413](https://github.com/grpc/grpc-go/runs/5744477531?check_suite_focus=true#step:8:413) +0x3128
  google.golang.org/grpc/internal/transport.NewClientTransport()
      /home/runner/work/grpc-go/grpc-go/internal/transport/transport.go:577 +0x386
  google.golang.org/grpc/internal/transport.setUpWithOptions()
      /home/runner/work/grpc-go/grpc-go/internal/transport/transport_test.go:453 +0x1f7
  google.golang.org/grpc/internal/transport.testFlowControlAccountCheck()
      /home/runner/work/grpc-go/grpc-go/internal/transport/transport_test.go:1498 +0x165
  google.golang.org/grpc/internal/transport.s.TestAccountCheckWindowSizeWithSmallWindow()
      /home/runner/work/grpc-go/grpc-go/internal/transport/transport_test.go:1478 +0x55
  google.golang.org/grpc/internal/transport.(*s).TestAccountCheckWindowSizeWithSmallWindow()
      <autogenerated>:1 +0x1d
  runtime.call16()
      /opt/hostedtoolcache/go/1.18.0/x64/src/runtime/asm_amd64.s:701 +0x48
  reflect.callMethod()
      <autogenerated>:1 +0x64
  reflect.methodValueCall()
      /opt/hostedtoolcache/go/1.18.0/x64/src/reflect/asm_amd64.s:76 +0x79
  testing.tRunner()
      /opt/hostedtoolcache/go/1.18.0/x64/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.18.0/x64/src/testing/testing.go:1486 +0x47
==================
--- FAIL: Test (107.41s)
    --- FAIL: Test/AccountCheckWindowSizeWithSmallWindow (10.06s)
        transport_test.go:138: Error on server while reading data header: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:138: Error on server while reading data header: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:145: Error on server while reading message: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:138: Error on server while reading data header: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:138: Error on server while reading data header: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:138: Error on server while reading data header: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:1550: Error on client while reading data: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:138: Error on server while reading data header: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:1550: Error on client while reading data: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:138: Error on server while reading data header: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:138: Error on server while reading data header: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:1550: Error on client while reading data: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:1544: Error on client while reading data frame header: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:1544: Error on client while reading data frame header: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:1550: Error on client while reading data: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:1550: Error on client while reading data: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:145: Error on server while reading message: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:1550: Error on client while reading data: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        transport_test.go:1580: Client expected an EOF from the server. Got: rpc error: code = DeadlineExceeded desc = context deadline exceeded
        testing.go:1312: race detected during execution of test
    testing.go:1312: race detected during execution of test
FAIL
FAIL	google.golang.org/grpc/internal/transport	208.[422](https://github.com/grpc/grpc-go/runs/5744477531?check_suite_focus=true#step:8:422)s
```

--- a/internal/transport/transport_test.go
+++ b/internal/transport/transport_test.go
@@ -1532,9 +1532,10 @@ func (s) TestAccountCheckWindowSizeWithLargeWindow(t *testing.T) {
 }
 
 func (s) TestAccountCheckWindowSizeWithSmallWindow(t *testing.T) {
+	// These settings disable dynamic window sizes based on BDP estimation;
+	// must be at least defaultWindowSize or the setting is ignored.
 	wc := windowSizeConfig{
 		serverStream: defaultWindowSize,
-		// Note this is smaller than initialConnWindowSize which is the current default.
 		serverConn:   defaultWindowSize,
 		clientStream: defaultWindowSize,
 		clientConn:   defaultWindowSize,
@@ -1576,10 +1577,11 @@ func testFlowControlAccountCheck(t *testing.T, msgSize int, wc windowSizeConfig)
 	for k := range server.conns {
 		st = k.(*http2Server)
 	}
+	server.mu.Unlock()
+
 	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	server.mu.Unlock()
-	const numStreams = 10
+	const numStreams = 5
 	clientStreams := make([]*Stream, numStreams)
 	for i := 0; i < numStreams; i++ {
 		var err error
@@ -1599,26 +1601,27 @@ func testFlowControlAccountCheck(t *testing.T, msgSize int, wc windowSizeConfig)
 			binary.BigEndian.PutUint32(buf[1:], uint32(msgSize))
 			opts := Options{}
 			header := make([]byte, 5)
-			for i := 1; i <= 10; i++ {
+			for i := 1; i <= 5; i++ {
 				if err := client.Write(stream, nil, buf, &opts); err != nil {
-					t.Errorf("Error on client while writing message: %v", err)
+					t.Errorf("Error on client while writing message %v on stream %v: %v", i, stream.id, err)
 					return
 				}
 				if _, err := stream.Read(header); err != nil {
-					t.Errorf("Error on client while reading data frame header: %v", err)
+					t.Errorf("Error on client while reading data frame header %v on stream %v: %v", i, stream.id, err)
 					return
 				}
 				sz := binary.BigEndian.Uint32(header[1:])
 				recvMsg := make([]byte, int(sz))
 				if _, err := stream.Read(recvMsg); err != nil {
-					t.Errorf("Error on client while reading data: %v", err)
+					t.Errorf("Error on client while reading data %v on stream %v: %v", i, stream.id, err)
 					return
 				}
 				if len(recvMsg) != msgSize {
-					t.Errorf("Length of message received by client: %v, want: %v", len(recvMsg), msgSize)
+					t.Errorf("Length of message %v received by client on stream %v: %v, want: %v", i, stream.id, len(recvMsg), msgSize)
 					return
 				}
 			}
+			t.Logf("stream %v done with pingpongs", stream.id)
 		}(stream)
 	}
 	wg.Wait()

